### DISCUSSION TITLE
HID upload for F0 variants

### DISCUSSION THREAD

**Discussion by geraldjust** [stm32duino/Arduino_Core_STM32#1448] [Q&A] [answered]
Hi, I saw from Here: https://github.com/stm32duino/Arduino_Core_STM32/issues/1000
you were able to crack a way to flash the firmware using the stock bootloader with a F070. I'm trying to use this: https://github.com/aroerina/STM32_HID_Bootloader_F0
which does the same thing as copying the Vector into ram since it does not have a VTOR register. it seems that i only had to modify the script.ld so make flash and sram to work per the example. Then copy over in main program before HAL_INIT(). I've been trying to get this to work with the stm32duino but have been running into a brick wall. Is there any way you can help me?  i got some pointers from the forum on were to place it but it does not seem to work. Weird thing is that it works just fine in cubeIDE.

**Comment by fpistm** [comment, marked as answer]
Hi @geraldjust 
Issue is for reporting error not request support. Moreover you already create a topic on this on the forum.
https://www.stm32duino.com/viewtopic.php?f=58&t=1180

**Comment by geraldjust** [comment]
Sorry though here i might be more promising to get someone who could help, Thanks!

**Comment by geraldjust** [comment]
So i did a couple of experiments. based on the malyan variant for the F070, i used `void initVariant()` so start the vector offset code for ram. I check that it worked by changing only ram location in the LD script and in the iniVariant BUT keeping the flash at 0x8000000. I uploaded it via DFU and a simple blinking worked perfectly. so i know the stm32duino is mapping over the ram as needed. But what i find is strange, when i try to change the Flash start position from 0x8000000 to 0x8001000 (in my example i expanded the bootloader to 4k) then its when it stops working. This is whats puzzling. Ram relocation works but not flash. I did change it both in the Ld script and the FLASH_VTABLE_ADDR that when inside 

`void initVariant()
{
  memcpy((void*)SRAM_VTABLE_ADDR,(void*)FLASH_VTABLE_ADDR,VTABLE_SIZE);
  SYSCFG->CFGR1 |= 3;
}`


This is in variant.cpp

**Comment by geraldjust** [comment]
found the problem(s)! first off with inside iniVariant() i added it closer / similar to my original post link. so this:

` volatile uint32_t *VectorTable = (volatile uint32_t *)0x20000000;
 uint32_t ui32_VectorIndex = 0;
 for(ui32_VectorIndex = 0; ui32_VectorIndex < 48; ui32_VectorIndex++)
 {
   VectorTable[ui32_VectorIndex] = *(__IO uint32_t*)((uint32_t)FLASH_VTABLE_ADDR + (ui32_VectorIndex << 2));
 }

  __HAL_RCC_USB_FORCE_RESET();
  __HAL_RCC_GPIOA_FORCE_RESET();

  __HAL_RCC_USB_RELEASE_RESET();
  __HAL_RCC_GPIOA_RELEASE_RESET();

      //  Enable SYSCFG peripheral clock
    __HAL_RCC_SYSCFG_CLK_ENABLE();
    __HAL_RCC_AHB_RELEASE_RESET();

  __HAL_SYSCFG_REMAPMEMORY_SYSTEMFLASH();
  __HAL_SYSCFG_REMAPMEMORY_SRAM();


__enable_irq();`

sysCFG clock must be on so i made sure it was in order for SYSCFG->CFGR1 to work. Then i reset the GPIO and the USB just in case, if the bootloader just finished flashing and jumps right into the program. and this finally make it work. 

Now the second problem is with the bootloader. since i was using it with a 128kB chip. It uses 2kb page size and the way its written it flashes the full 2k at once. So even if you use the HID-flash program, your program size rounded to to the nearest 1024 bytes. This gave a 50/50 chance that it will flash the whole file or cut it off la last bytes. I will go head and try this from now on with a 64k chip to keep it simple with a 1k page size. 

Anyways. i think adding this in `initVariant ` would not be the best solution. But it works for now.
